scheduler/servicestatemanager/servicestatemanager.go

```diff
@@ -269,6 +289,112 @@ func (s *BatchServiceStateManager) removeTenant(tenantID string) error {
 	return nil
 }
 
+func (s *BatchServiceStateManager) SyncCurrentStates(svcIDs []string) {
+	states := make(map[string]service.ServiceCurrentState)
+	s.RLock()
+	defer s.RUnlock()
+
+	// Get states for everything in our queues
+	for _, tenantQueues := range s.TenantQueues {
+		for _, queue := range tenantQueues {
+			func() {
+				queue.RLock()
+				defer queue.RUnlock()
+
+				// Pending states for services in the queue
+				for _, batch := range queue.BatchQueue {
+					pendingState := service.DesiredToCurrentPendingState(batch.DesiredState, batch.Emergency)
+					for _, svc := range batch.Services {
+						states[svc.ID] = pendingState
+					}
+				}
+
+				// Transition states for services in the current batch
+				transitionState := service.DesiredToCurrentTransitionState(queue.CurrentBatch.DesiredState, queue.CurrentBatch.Emergency)
+				for _, svc := range queue.CurrentBatch.Services {
+					states[svc.ID] = transitionState
+				}
+			}()
+		}
+	}
+
+	// Update the states we know and build a list of missing states
+	s.currentStateLock.Lock()
+	defer s.currentStateLock.Unlock()
+
+	var missingStates []string
+	for _, sid := range svcIDs {
+		if state, ok := states[sid]; ok {
+			s.updateServiceCurrentState(state, sid)
+		} else {
+			missingStates = append(missingStates, sid)
+		}
+	}
+
+	// Now figure out the ones that aren't in the queues
+	// Get updated service info
+	svcs := s.Facade.GetServicesForScheduling(s.ctx, missingStates)
+	for _, svc := range svcs {
+		s.startCurrentStateWait(svc, service.DesiredState(svc.DesiredState), svc.EmergencyShutdown)
+	}
+}
+
+func (s *BatchServiceStateManager) startCurrentStateWait(svc *service.Service, desiredState service.DesiredState, emergency bool) {
+	if thread, ok := s.currentStateWaits[svc.ID]; ok {
+		if thread.WaitingState == desiredState {
+			// We are already waiting on it, do nothing
+			return
+		} else {
+			// Cancel this wait and start a new one
+			close(thread.Cancel)
+			<-thread.Done
+			delete(s.currentStateWaits, svc.ID)
+		}
+	}
+
+	finalState := service.DesiredToCurrentFinalState(desiredState, emergency)
+	currentState := service.DesiredToCurrentTransitionState(desiredState, emergency)
+
+	// Set current state to the transition state
+	s.updateServiceCurrentState(currentState, svc.ID)
```

Comment:
The calling method should (and does) get the lock.  I added a comment on both methods stating that the lock is required before calling.